server/src/main/java/org/elasticsearch/cluster/AckedClusterStateUpdateTaskChain.java

```diff
@@ -0,0 +1,132 @@
+
+    private static final Logger LOGGER = LogManager.getLogger(AckedClusterStateUpdateTaskChain.class);
+
+    private final ChainableAckedClusterStateUpdateTask firstTask;
+    private final ClusterService clusterService;
+
+    public AckedClusterStateUpdateTaskChain(List<ChainableTask> taskChain,
+                                            ClusterService clusterService,
+                                            ActionListener<AcknowledgedResponse> listener,
+                                            AcknowledgedRequest request,
+                                            Priority priority) {
+        Objects.requireNonNull(taskChain, "Task chain cannot be null.");
+        assert !taskChain.isEmpty() : "Task chain must consist of at least one task.";
+        this.clusterService = clusterService;
+
+        // Wrap into next task aware instances.
+        ChainableAckedClusterStateUpdateTask current = null;
+        ChainableAckedClusterStateUpdateTask next = null;
+        ListIterator<ChainableTask> it = taskChain.listIterator(taskChain.size());
+        while (it.hasPrevious()) {
+            current = new ChainableAckedClusterStateUpdateTask(priority, request, listener, it.previous(), next);
+            next = current; // We are going from right to left, preparing next = "i+1", for the next iteration.
+        }
+        this.firstTask = current;
+    }
+
+
+    public static record ChainableTask(Function<ClusterState, ClusterState> stateUpdater, String hintOnError, String taskName) { }
+
+    public void startTaskChain() {
+        clusterService.submitStateUpdateTask(firstTask.taskName, firstTask);
+    }
+
+    public final class ChainableAckedClusterStateUpdateTask extends AckedClusterStateUpdateTask {
+
+        private final ActionListener listener;
+        private final ChainableAckedClusterStateUpdateTask nextTask;
+        private final Function<ClusterState, ClusterState> stateUpdater;
+        private final String hintOnError;
+        private final String taskName;
+
+        public ChainableAckedClusterStateUpdateTask(Priority priority,
+                                                     AckedRequest request,
+                                                     ActionListener listener,
+                                                     ChainableTask chainableTask,
+                                                     @Nullable ChainableAckedClusterStateUpdateTask nextTask) {
+            super(priority, request, listener);
+            this.listener = listener;
+            this.stateUpdater = chainableTask.stateUpdater;
+            this.taskName = chainableTask.taskName;
+            this.hintOnError = chainableTask.hintOnError;
+            this.nextTask = nextTask;
+        }
+
+        @Override
+        public void onAllNodesAcked(@Nullable Exception e) {
+            if (e == null) {
+                /*
+                if (!clusterService.state().nodes().isLocalNodeElectedMaster()) {
```

Comment:
not sure about that, just noticed such logic in https://github.com/crate/crate/blob/d872104f5b124a54c2725044c45696ea7196446d/server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java#L133-L136